src/service_discovery/mod.rs

```diff
@@ -105,35 +111,40 @@ impl ServiceDiscovery {
     }
 
     fn read(&mut self, core: &mut Core, poll: &Poll) {
-        let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
-            Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
-            // TODO(povilas): handle WouldBlock as well
-            Err(ref e) if e.kind() == ErrorKind::Interrupted => return,
-            Err(e) => {
-                debug!("ServiceDiscovery error in read: {:?}", e);
-                self.terminate(core, poll);
-                return;
-            }
-        };
-
-        let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
-            Ok(msg) => msg,
-            Err(e) => {
-                debug!("Bogus message serialisation error: {:?}", e);
-                return;
-            }
-        };
+        loop {
+            let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
+                Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
+                Err(ref e)
+                    if e.kind() == ErrorKind::Interrupted || e.kind() == ErrorKind::WouldBlock =>
+                {
+                    return
+                }
+                Err(e) => {
+                    debug!("ServiceDiscovery error in read: {:?}", e);
+                    self.terminate(core, poll);
+                    return;
+                }
+            };
 
-        match msg {
-            DiscoveryMsg::Request { guid } => {
-                if self.listen && self.guid != guid {
-                    self.reply_to.push_back(peer_addr);
-                    self.write(core, poll)
+            let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
```

Comment:
maybe do this inline in the match arm of `recv_from` above.
This will ensure that
1. you don't discard all (from potentially various peers) but the last message
1. would blocks/interrupts after reading a valid message don't make you discard the messages read so far in this run of the loop